### PR TITLE
polyphone_swiss_french recipe updates

### PR DESCRIPTION
**Pull request by AdolfVonKleist**
This pull request aims to update the Polyphone Swiss French recipe. It proposes:

- fix to the `local/data_prep` script adding `.py` extension and fixes to subsequent linting complaints
- fix to reference text encoding detection to deal with windows based encoding oddities 
- new egs transformer results
- first egs conformer results
- first egs transducer results
- addition of first egs2 conformer results

This is a first attempt at using the latest transducer-related updates.  I still get a slightly better result with the conformer in both egs and egs2.  Perhaps room for improvement there in a future update; I only took one stab at it.

All the new results were run on an 8x NVIDIA Quadro RTX 5000 @ 16GB.

The conformer results for egs2 use the default speed perturbation as well.

I believe can also upload the packed model for egs2 if there is an interest.

### CODE REVIEW

**Comment by b-flo** [comment]
Hi,

Thank you for the contribution!

> This is a first attempt at using the latest transducer-related updates.

When you say latest updates, you're talking in regards to `master` or #2727 ? If it's the former, some fixes related to rel. pos. enc. and VGG2L are missing, sorry...

**Comment by joespitch** [comment]
@b-flo thanks for your feedback.  In fact, for the transducer I used the 'bleeding' edge from #2727 and it should leverage everything up to https://github.com/espnet/espnet/pull/2727/commits/f3ccb2255c86cd838cb1d6da62bf541d29e8b06a (second to last commit?). However I have not made any attempt yet at tuning the transducer setup; this was just a 'make it work' first attempt with a slightly larger dataset.  There are still only a small number of extant egs and I think maybe none so far(?) with the larger datasets; but this seemed like a good start.  I'll try the suggestions for the decoder config as well.

**Comment by AdolfVonKleist** [comment]
@b-flo  thanks for your feedback. In fact, for the transducer I used the 'bleeding' edge from #2727 and it should leverage everything up to f3ccb22 (second to last commit?). However I have not made any attempt yet at tuning the transducer setup; this was just a 'make it work' first attempt with a slightly larger dataset. There are still only a small number of extant egs and I think maybe none so far(?) with the larger datasets; but this seemed like a good start. I'll try the suggestions for the decoder config as well.

**Comment by b-flo** [comment]
> However I have not made any attempt yet at tuning the transducer setup; this was just a 'make it work' first attempt with a slightly larger dataset.

Yes, that's why I only proposed testing some other decoding settings for now. It's a bit difficult to comment on tuning others aspects (architecture, training) without additional informations! ;)

>  There are still only a small number of extant egs and I think maybe none so far(?) with the larger datasets; but this seemed like a good start. I'll try the suggestions for the decoder config as well.

We did train models (RNN-T, T-T) on larger dataset such as Librispeech (960 hours) and ESTER (300 hours) some time ago but it was not added to ESPnet.
Also, we'll accelerate works related to transducer in the following months, adding transducer configs for most recipe being one of the main items.

**Comment by AdolfVonKleist** [comment]
@b-flo I ran a few more iterations of the evaluation over night.  In the end it did yield a modest improvement over the first baseline (`CORR dev: 93.5, CORR test: 93.3`) to:
 * `n_average: 10 -> 20`
 * `lm: 0.3 -> 0.1`
 * `beam: 10 -> 20`
--------------------
```
 ### WER
|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_valid_dev_decode_lm|9846|72368|93.8|5.4|0.9|0.9|7.2|23.5|
|decode_valid_test_decode_lm|4261|31446|93.5|5.5|1.0|1.0|7.5|23.8|
```
however I would 'guesstimate' that there is more to be gained from tweaking the architecture at this point. I am moving all attention into `espnet2` for future development, but would of course be curious about the outcomes you obtained from the Librispeech and ESTER experiments you mention.  Did you achieve an improvement over the general offline conformer?  Are you also looking at other factors like RTF vs WAcc tradeoff (I know this is not a major focus for the project, but it is definitely relevant to many use cases).  Thanks again.  I'll update the pull request with those last figures and then probably call it a day on this one unless there are some additional requests.

**Comment by b-flo** [comment]
> however I would 'guesstimate' that there is more to be gained from tweaking the architecture at this point. 

Yes but it was faster to modify decoding right now than asking you to test things for training without having seen the different logs ;)

> I am moving all attention into espnet2 for future development, but would of course be curious about the outcomes you obtained from the Librispeech and ESTER experiments you mention.

Note that transducer porting to ESPnet2 is in the pipeline but not available yet, a few things need to be discussed before merging.

> but would of course be curious about the outcomes you obtained from the Librispeech and ESTER experiments you mention.  Did you achieve an improvement over the general offline conformer?

Sorry we didn't experiment with conformer for these ones yet.
For Librispeech, we got a WER of 3.0/6.9 following [Google's paper](https://arxiv.org/abs/2002.02562) but there are some differences in training and architecture we can do in ESPnet. For ESTER, we only compared RNN architectures yet, experiments for transformer and conformer are on the way.

> Are you also looking at other factors like RTF vs WAcc tradeoff (I know this is not a major focus for the project, but it is definitely relevant to many use cases).

Depending on the dataset, you can use decoding algorithm such as NSC or TSD with a low parameters (`nstep: 1`, `max_sym_exp`: 2) to achieve comparable or better performances than default beam search (being CER/WER or RTF).

It was indeed not a focus for the project but we're now working on reducing memory, accelerating training and inference, and also streaming models. I can give an ETA but we're aiming at start/mid spring.

**Comment by joespitch** [comment]
Cool. Ok last inquiry for the moment: is anyone looking at [Universal ASR](https://arxiv.org/pdf/2010.06030.pdf)? This is another Google paper where they achieve some further recent improvements over the Zhang approach in terms of latency, flexibility and general accuracy.

**Comment by AdolfVonKleist** [comment]
Ok cool thanks for sharing [I keep context switching...].  Last enquiry then: is anyone currently looking at [Universal ASR](https://arxiv.org/pdf/2010.06030.pdf)? This is another recent Google result which achieves some further improvements on both the latency and general accuracy fronts.  The dual-mode training aspect is particularly appealing.

**Comment by b-flo** [comment]
>  Last enquiry then: is anyone currently looking at Universal ASR?

Not that I'm aware of, you can open a new issue to request the feature though! Who knows, we may consider it when working on streaming models.

**Comment by AdolfVonKleist** [comment]
Actually I meant more, would there be an interest in such a contribution? I don't know that I will actually have the time to work on it in the next months, but would consider looking at it if there is not already ongoing work in that direction.  I see questions around this subject asked frequently, but the answers tend to be somewhat variable.

**Comment by b-flo** [comment]
> Actually I meant more, would there be an interest in such a contribution? 

For sure, I don't think there are work in progress for that. @sw005320 can you confirm?

**Comment by sw005320** [comment]
> Ok cool thanks for sharing [I keep context switching...]. Last enquiry then: is anyone currently looking at [Universal ASR](https://arxiv.org/pdf/2010.06030.pdf)? This is another recent Google result which achieves some further improvements over the Zhang approach on both the latency and general accuracy fronts. The dual-mode training aspect is particularly appealing.

Sorry for my late chime in.
Yes, we have a couple of trials internally.

**Comment by sw005320** [comment]
> > Actually I meant more, would there be an interest in such a contribution?
> 
> For sure, I don't think there are work in progress for that. @sw005320 can you confirm?

This is really great.
We just launched an internal RNN-T/streaming project. I'd like you to be involved in the project. (it's not commitment. If you have time and interest, you can just contribute).

I'll send an invitation to our lab's slack.

File: egs2/polyphone_swiss_french/asr1/run.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+train_set="valid_train"
+valid_set="valid_dev"
+test_sets="valid_test"
```
Comment (sw005320) [inline egs2/polyphone_swiss_french/asr1/run.sh:10]:
```suggestion
test_sets="valid_dev valid_test"
```
`test_sets` do not include the validation set (the validation set is only used during training). If you want to have ASR results in the validation set, you can specify like that.

**Review by sw005320 — COMMENTED** [review]
Can you upload a model as well?

File: egs2/polyphone_swiss_french/asr1/run.sh (hunk 1 above)
Comment (joespitch) [reply egs2/polyphone_swiss_french/asr1/run.sh:10]: Ah ok I did not check this closely enough and copied the content from egs1.  Yes I can upload the egs2 model.

Comment (AdolfVonKleist) [reply egs2/polyphone_swiss_french/asr1/run.sh:10]: Thanks! I copied the egs1 content without checking this change closely enough.  Yes, I can upload the egs2 model.

Comment (AdolfVonKleist) [reply egs2/polyphone_swiss_french/asr1/run.sh]: I've added the egs2 reference model, as well as updated the results with the dev evaluation.  Not clear to me what the codecov issue is as this update makes no material changes to any part of the espnet core, and only only makes your recommended commit suggestion to `run.sh`, adding `valid_dev`.

**Comment by sw005320** [comment]
Many thanks, @AdolfVonKleist!!!

**Pull request merged**